utils/vptr/include/virtual_ptr.hpp

```diff
@@ -0,0 +1,510 @@
+      // with the size of the current one
+      auto prev_node = std::prev(node);
+      if (!prev_node->second._free) {
+        break;
+      }
+      prev_node->second._size += node->second._size;
+
+      // remove the current node
+      m_freeList.erase(node);
+      m_pointerMap.erase(node);
+
+      // point to the previous node
+      node = prev_node;
+    }
+  }
+
+  /* remove_pointer.
+   * Removes the given pointer from the map.
+   */
+  void remove_pointer(const virtual_pointer_t ptr) {
+    auto node = this->get_node(ptr);
+
+    node->second._free = true;
+    m_freeList.emplace(node);
+
+    // Fuse the node
+    // with free nodes before and after it
+    fuse_forward(node);
+    fuse_backward(node);
+
+    // If after fusing the node is the last one
+    // simply remove it (since it is free)
+    if (node == std::prev(m_pointerMap.end())) {
+      m_freeList.erase(node);
+      m_pointerMap.erase(node);
+    }
+
+#if VIRTUAL_PTR_VERBOSE
+    std::cout << "New list after removing: " << static_cast<long>(ptr)
+              << std::endl;
+    for (auto &n : m_pointerMap) {
+      std::cout << static_cast<long>(n.first) << " { "
+                << n.second._b.get_impl().get() << ", "
+                << ((n.second._free) ? "Freed" : "Usable") << ", "
+                << n.second._b.get_count() << ", " << n.second._size << " }"
+                << std::endl;
+    }
+#endif  // VIRTUAL_PTR_VERBOSE
+  }
+
+  /* count.
+   * Return the number of active pointers (i.e, pointers that
+   * have been malloc but not freed).
+   */
+  size_t count() const {
+    if (!m_pointerMap.empty()) {
+#if VIRTUAL_PTR_VERBOSE
+      std::cout << " Map size " << m_pointerMap.size() << " "
+                << m_freeList.size() << std::endl;
+#endif  // VERBOSE
+      return (m_pointerMap.size() - m_freeList.size());
+    } else {
+      return 0;
+    }
+  }
+
+ private:
+  /* Maps the pointer addresses to buffer and size pairs.
+    */
+  pointerMap_t m_pointerMap;
+
+  /**
+   * Compare two iterators to pointer map entries according to
+   * the size of the allocation on the device.
+   */
+  struct SortBySize {
+    bool operator()(typename pointerMap_t::iterator a,
+                    typename pointerMap_t::iterator b) {
+      return ((a->first < b->first) && (a->second <= b->second)) ||
+             ((a->first < b->first) && (b->second <= a->second));
+    }
+  };
+
+  /* List of free nodes available for re-using
+   */
+  std::set<typename pointerMap_t::iterator, SortBySize> m_freeList;
+};
+
+/**
+ * Malloc-like interface to the pointer-mapper.
+ * Given a size, creates a byte-typed buffer and returns a
+ * fake pointer to keep track of it.
+ * \param size Size in bytes of the desired allocation
+ * \throw cl::sycl::exception if error while creating the buffer
+ */
+template <
+    typename buffer_allocator = cl::sycl::default_allocator<buffer_data_type> >
+inline void *SYCLmalloc(size_t size, PointerMapper &pMap) {
```

Comment:
It looks to me like you don't actually need inline, but honestly, it is not a problem to have it and I am going to merge this today!